ccp/src/hashrate/hashrate_collector.rs

```diff
@@ -0,0 +1,240 @@
+                self.new_epoch(hashrate_record.epoch);
+
+                Some(statistics)
+            }
+        };
+
+        match self.entries.entry(hashrate_record.core_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(CUHashrateRaw::from_single_record(hashrate_record));
+            }
+            Entry::Occupied(entry) => entry.into_mut().count_record(hashrate_record),
+        };
+
+        result
+    }
+
+    pub(crate) fn collect(&self) -> HashMap<LogicalCoreId, CUHashrate> {
+        let epoch_duration = match self.status {
+            CollectorStatus::Started { started_time, .. } => started_time.elapsed(),
+            CollectorStatus::JustCreated => return HashMap::new(),
+        };
+
+        let epoch_duration = epoch_duration.as_secs_f64();
+
+        self.entries
+            .iter()
+            .map(|(&core_id, info)| {
+                let hashrate = match info.pow_duration {
+                    ParameterStatus::Measured(duration) => {
+                        let hashrate = info.hashes_checked as f64 / duration.as_secs_f64();
+                        ParameterStatus::Measured(hashrate)
+                    }
+                    ParameterStatus::NotMeasured => ParameterStatus::NotMeasured,
+                };
+
+                let dataset_initialization = match info.dataset_initialization.iter().max() {
+                    Some(&duration) => ParameterStatus::Measured(duration),
+                    None => ParameterStatus::NotMeasured,
+                };
+
+                let statistics = CUHashrate {
+                    effective_hashrate: info.hashes_checked as f64 / epoch_duration,
+                    hashrate,
+                    cache_creation: info.cache_creation,
+                    dataset_initialization,
+                };
+
+                (core_id, statistics)
+            })
+            .collect::<HashMap<_, _>>()
+    }
+
+    pub(crate) fn proof_found(&mut self, core_id: LogicalCoreId) {
+        use std::collections::hash_map::Entry;
+
+        match self.entries.entry(core_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(CUHashrateRaw::from_proof_found());
+            }
+            Entry::Occupied(entry) => entry.into_mut().count_proof_found(),
+        };
+    }
+
+    fn new_epoch(&mut self, epoch: EpochParameters) {
+        self.status = CollectorStatus::Started {
+            started_time: Instant::now(),
+            epoch,
+        };
+
+        self.entries.clear();
+    }
+}
+
+impl Default for CUHashrateRaw {
+    fn default() -> Self {
+        Self {
+            cache_creation: ParameterStatus::NotMeasured,
+            dataset_initialization: vec![],
+            pow_duration: ParameterStatus::NotMeasured,
+            hashes_checked: 0,
+            proofs_found: 0,
+        }
+    }
+}
+
+impl CUHashrateRaw {
+    pub(self) fn from_single_record(entry: HashrateCURecord) -> Self {
```

Comment:
Is it different from `CUHashrateRaw::default().count_record(entry)`?

If so, the code may be shorter, and entry API usage above can be simplified.
If not, it should be commented.